update-engine/src/events.rs

```diff
@@ -0,0 +1,688 @@
+        outcome: StepOutcome<S>,
+
+        /// The next step that is being started.
+        next_step: StepInfoWithMetadata<S>,
+
+        /// Total time elapsed since the start of the step. Includes prior
+        /// attempts.
+        step_elapsed: Duration,
+
+        /// The time it took for this attempt to complete.
+        attempt_elapsed: Duration,
+    },
+
+    /// Execution is complete.
+    ///
+    /// This is a terminal event: it is guaranteed that no more events will be
+    /// seen after this one.
+    ExecutionCompleted {
+        /// Information about the last step that completed.
+        last_step: StepInfoWithMetadata<S>,
+
+        /// The attempt number that completed.
+        last_attempt: usize,
+
+        /// The outcome of the last step.
+        last_outcome: StepOutcome<S>,
+
+        /// Total time elapsed since the start of the step. Includes prior
+        /// attempts.
+        step_elapsed: Duration,
+
+        /// The time it took for this attempt to complete.
+        attempt_elapsed: Duration,
+    },
+
+    /// Execution failed.
+    ///
+    /// This is a terminal event: it is guaranteed that no more events will be
+    /// seen after this one.
+    ExecutionFailed {
+        /// Information about the step that failed.
+        failed_step: StepInfoWithMetadata<S>,
+
+        /// The total number of attempts that were performed before the step failed.
+        total_attempts: usize,
+
+        /// Total time elapsed since the start of the step. Includes prior
+        /// attempts.
+        step_elapsed: Duration,
+
+        /// The time it took for this attempt to complete.
+        attempt_elapsed: Duration,
+
+        /// A message associated with the failure.
+        message: String,
+
+        /// A chain of causes associated with the failure.
+        causes: Vec<String>,
+    },
+
+    /// Future variants that might be unknown.
+    #[serde(other, deserialize_with = "deserialize_ignore_any")]
+    Unknown,
+}
+
+impl<S: StepSpec> StepEventKind<S> {
+    /// Returns the priority of the event.
+    ///
+    /// For more about this, see [`StepEventPriority`].
+    pub fn priority(&self) -> StepEventPriority {
+        match self {
+            StepEventKind::NoStepsDefined
+            | StepEventKind::ExecutionStarted { .. }
+            | StepEventKind::StepCompleted { .. }
+            | StepEventKind::ExecutionCompleted { .. }
+            | StepEventKind::ExecutionFailed { .. } => StepEventPriority::High,
+            StepEventKind::ProgressReset { .. }
+            | StepEventKind::AttemptRetry { .. }
+            | StepEventKind::Unknown => StepEventPriority::Low,
+        }
+    }
+}
+
+/// The priority of a [`StepEvent`].
+///
+/// Returned by [`StepEventKind::priority`].
+///
+/// Some [`StepEvent`]s have a higher priority than others. For example, events
+/// related to step successes and failures must be delivered, while events
+/// related to retries can be dropped.
```

Comment:
Maybe worth expanding "can be dropped" a bit. Is this to help installinator filter down the event list to fit under its cap for sending within one message?

Reply:
Yeah, it's specifically for that purpose.

Reply:
Expanded a bit.